### PR TITLE
Add stable nrm2 Reducer

### PR DESCRIPTION
**Pull request by leezu**
## Description ##
The mxnet L2 norm implementation is currently prone to under/overflow when called on NDArrays with very small or large elements. This PR adds a stable L2 norm reducer and switches the dense `norm` operator to make use of it.

The implementation follows the [Blas nrm2 reference implementation](http://www.netlib.org/blas/snrm2.f).

As  a final rescaling step is necessary  at the end of the reduction, this PR also introduces a new `Finalize` method for Reducers. This method is a noop for all previous reducers. A separate PR needs to add the `Finalize` method to the maximum and minimum reducers in mshadow.

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [X] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [X] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [x] Improve numerical stability for L2 norm in `norm()` operator
- [x] Improve numerical stability of sum in broadcast reduce

### Comments ###
- The test_norm() test was currently disabled (https://github.com/apache/incubator-mxnet/issues/11509). It is enabled in this PR. Maybe using a different norm2 implementation will fix the flakiness of the test..
- I tried avoiding the need to add the `Finalize` method to all existing reducers (i.e. maximum, minimum, sum, product, nansum, nanprod) using [SFINAE](https://en.cppreference.com/w/cpp/language/sfinae) but it didn't get it to work immediately (the Finalize() would correctly be skipped if it the respective reducer didn't implement it, but for unknown reasons was also skipped if the reducer implemented it). As there are not many reducers, I added the method for now. The SFINAE approach  was as  follows:

``` C++
template <typename Reducer, typename DType>
auto reducer_finalize(DType val, DType residual, int) -> decltype(Reducer::Finalize, void()) {
   Reducer::Finalize(val, residual);
}

template <typename Reducer, typename DType>
auto reducer_finalize(DType val, DType residual, long) -> decltype(void()) {}
```

### CODE REVIEW

**Comment by leezu** [comment]
@anirudh2290 @szha This is ready for review. [Tests are passing](jenkins.mxnet-ci.amazon-ml.com/blue/organizations/jenkins/incubator-mxnet/detail/PR-11573/10/pipeline), but for some reason the deploy stage failed due to a Jenkins error. Rerunning now.

File: tests/python/unittest/test_operator.py [hunk 1]
```diff
@@ -3051,30 +3061,31 @@ def l2norm(input_data, axis=0, keepdims=True):
...
-                                           atol=1e-2 if dtype is np.float16 else 1e-5, ctx=ctx)
-                    check_symbolic_backward(norm_sym, [in_data], [np.ones(npy_out.shape)],
-                                            [npy_out_backward],
-                                            rtol=1e-2 if dtype is np.float16 else 1e-5,
-                                            atol=1e-2 if dtype is np.float16 else 1e-5, ctx=ctx)
-                    # check gradient
-                    check_numeric_gradient(norm_sym, [in_data], numeric_eps=epsilon, rtol=1e-2, atol=1e-3)
-                        
+                # Disable numeric gradient https://github.com/apache/incubator-mxnet/issues/11509
+                # # check gradient
```
Comment (anirudh2290) [inline tests/python/unittest/test_operator.py]: Is this supposed to be commented. Did this not help for #11509 ?

File: src/operator/mshadow_op.h [hunk 2]
```diff
@@ -815,6 +868,76 @@ struct nanprod {
...
+  /*! \brief combine the results of two reducers */
+  template<typename DType>
+  MSHADOW_XINLINE static void Merge(volatile DType& dst_val, volatile DType& src_val) { // NOLINT(*)
+    dst_val += src_val;
+  }
+  /*! \brief combine the results of two reducers */
+  template<typename DType>
+  MSHADOW_XINLINE static void Merge(volatile DType& dst_ssq, volatile DType& dst_scale, volatile DType& src_ssq, volatile DType& src_scale) { // NOLINT(*)
+    if (dst_scale != 0 && dst_scale >= src_scale) {
+      dst_ssq = dst_ssq + src_ssq * (src_scale / dst_scale) * (src_scale / dst_scale);
```
Comment (anirudh2290) [inline src/operator/mshadow_op.h]: can you please elaborate on how this expression was obtained

File: tests/python/unittest/test_operator.py (hunk 1 above)
Comment (leezu) [reply tests/python/unittest/test_operator.py]: Unfortunately it didn't help. I observed a similar error to https://github.com/apache/incubator-mxnet/issues/11509#issuecomment-401543334

File: src/operator/mshadow_op.h (hunk 2 above)
Comment (anirudh2290) [reply src/operator/mshadow_op.h]: thanks for the explanation!

**Pull request merged**